keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+	metav1.TypeMeta   `json:",inline"`
+
+	Spec ClaimParameterSpec `json:"spec"`
+}
+
+type ClaimParameterSpec struct {
+	TypeMeta `json:",inline"` // Not metav1.TypeMeta! See below.
+
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=1
+    // +kubebuilder:validation:Maximum=16
+	Count int64 `json:"count,omitempty"`
+
+	Env map[string]string `json:"env,omitempty"`
+}
+
+type TypeMeta struct {
+	// +kubebuilder:default=Capacity
+	// +kubebuilder:validation:Enum=Capacity
+	Kind string `json:"kind,omitempty"`
+
+	// +kubebuilder:default=counter.dra.config.k8s.io/v1alpha1
+	// +kubebuilder:validation:Enum=counter.dra.config.k8s.io/v1alpha1
+	APIVersion string `json:"apiVersion,omitempty"`
+}
+```
+
+In this example, the driver supports the `Count` numeric parameter and adds
+environment variables as configuration parameters. There is a 1:1 relationship
+between the CRD type and the numeric type that it is using. Users define
+parameters like they would without numeric parameters:
+
+```
+kind: ClaimParameter
+apiVersion: dra.e2e.example.com
+spec:
+  count: 10
+  env:
+    HELLO_WORLD: "Thanks for the fish."
+```
+
+### Notes/Constraints/Caveats
+
+For this proposal to work, kube-scheduler must have permission to read, list and
+watch arbitrary resources. This is necessary because it is not known in advance
+which CRDs will be used by DRA drivers.
+
+This may change at runtime as classes get added or removed. To avoid repeated
+GET operations for those CRDs, dynamic informers will have to be used that get
+started and stopped as needed.
+
+### Risks and Mitigations
+
+Granting kube-scheduler permission to read arbitrary types could be used by an
+attacker to extract information. However, for that an attacker first would have
+to gain control of the scheduler, get it to execute API calls, and then
+retrieve the result. This seems like a low risk.
+
+Using a single NodeResourceCapacity object per node is simple and convenient
+also for admins, but might run into size limitations. This should not be a
+problem in practice because the number of drivers per node is expected to be
+low and the capacity information from each should be small. Limits for the
+JSON fragment can be introduced to prevent excessive sizes.
+
+## Design Details
+
+### Decoding claim parameters
+
+Extracting the numeric parameters can be done without knowing the CRD:
+
+- Look up the resource based on group and kind with a REST mapper.
+
+- Retrieve the object with a `dynamic.Client` as an `unstructured.Unstructured`
+  instance (basically a nested map).
+
+- Retrieve the sub-map via the JSON path.
+
+- Re-encode that part as JSON with the normal `encode/json` package.  In the
+  example above, this will include `kind` and `apiVersion:
+  counter.dra.config.k8s.io` because those are the defaults and only allowed
+  values.
+
+- Decode into the numeric parameter with the apimachinery JSON decoder,
+  ignoring unknown fields like `env` in this example.
```

Comment:
This goes back to my previous question...

Just to make sure I understand, this implies that a `ClaimParameters` object *MUST* be a superset of the number `Parameters` type defined for this resource. Where anything *not* included in the `Parameters` type is assumed to be interpreted by the kubelet plugin when actually preparing an allocated claim.

Reply:
Your understanding is correct.

There are some alternatives for the process for extracting numeric parameters from the vendor CRD:
- field path (this text here)
- Go templates (https://github.com/kubernetes/enhancements/pull/4384/files#r1455827364)
- CEL (https://github.com/kubernetes/enhancements/pull/4384/files#r1467088455)

John and I prefer CEL, but I need to try it out first. The concept remains the same (vendor CRD -> numeric model's parameter type).